### PR TITLE
Updated `volumes-web-app`

### PR DESCRIPTION
**Pull request by TakoB222**
### Description
---
- Rolled back source-subdir
- Added tests
- Added tox.ini

### CODE REVIEW

**Comment by kimwnasptd** [comment]
@TakoB222 can you ensure the `rockcraft.yaml` aligns with the structure for the other similar web app rocks?
https://github.com/canonical/kubeflow-rocks/blob/main/jupyter-web-app/rockcraft.yaml

**Comment by TakoB222** [comment]
> @TakoB222 can you ensure the `rockcraft.yaml` aligns with the structure for the other similar web app rocks? https://github.com/canonical/kubeflow-rocks/blob/main/jupyter-web-app/rockcraft.yaml

@kimwnasptd updated

**Comment by kimwnasptd** [comment]
@TakoB222 when I try to run the image locally I'm getting the following errors:

```bash
rockcraft pack
ROCK=volumes-web-app_v1.9.0_amd64.rock
IMG=kimwnasptd/vwa:rock-1
sudo rockcraft.skopeo --insecure-policy \
	copy \
	oci-archive:$ROCK \
	docker-daemon:$IMG

docker run -ti $IMG -v
2024-08-13T17:36:11.514Z [pebble] Started daemon.
2024-08-13T17:36:11.521Z [pebble] POST /v1/services 3.441549ms 202
2024-08-13T17:36:11.525Z [pebble] Service "volumes-ui" starting: /bin/bash -c "gunicorn -w 3 --bind 0.0.0.0:5000 --access-logfile - entrypoint:app"
2024-08-13T17:36:11.720Z [volumes-ui] [2024-08-13 17:36:11 +0000] [15] [INFO] Starting gunicorn 23.0.0
2024-08-13T17:36:11.720Z [volumes-ui] [2024-08-13 17:36:11 +0000] [15] [INFO] Listening at: http://0.0.0.0:5000 (15)
2024-08-13T17:36:11.720Z [volumes-ui] [2024-08-13 17:36:11 +0000] [15] [INFO] Using worker: sync
2024-08-13T17:36:11.723Z [volumes-ui] [2024-08-13 17:36:11 +0000] [17] [INFO] Booting worker with pid: 17
2024-08-13T17:36:11.732Z [volumes-ui] [2024-08-13 17:36:11 +0000] [17] [ERROR] Exception in worker process
2024-08-13T17:36:11.732Z [volumes-ui] Traceback (most recent call last):
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/arbiter.py", line 608, in spawn_worker
2024-08-13T17:36:11.732Z [volumes-ui]     worker.init_process()
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/workers/base.py", line 135, in init_process
2024-08-13T17:36:11.732Z [volumes-ui]     self.load_wsgi()
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/workers/base.py", line 147, in load_wsgi
2024-08-13T17:36:11.732Z [volumes-ui]     self.wsgi = self.app.wsgi()
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/app/base.py", line 66, in wsgi
2024-08-13T17:36:11.732Z [volumes-ui]     self.callable = self.load()
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 57, in load
2024-08-13T17:36:11.732Z [volumes-ui]     return self.load_wsgiapp()
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/app/wsgiapp.py", line 47, in load_wsgiapp
2024-08-13T17:36:11.732Z [volumes-ui]     return util.import_app(self.app_uri)
2024-08-13T17:36:11.732Z [volumes-ui]   File "/lib/python3.8/site-packages/gunicorn/util.py", line 370, in import_app
2024-08-13T17:36:11.732Z [volumes-ui]     mod = importlib.import_module(module)
2024-08-13T17:36:11.732Z [volumes-ui]   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
2024-08-13T17:36:11.732Z [volumes-ui]     return _bootstrap._gcd_import(name[level:], package, level)
2024-08-13T17:36:11.732Z [volumes-ui]   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
2024-08-13T17:36:11.732Z [volumes-ui]   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
2024-08-13T17:36:11.732Z [volumes-ui]   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
2024-08-13T17:36:11.732Z [volumes-ui]   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
2024-08-13T17:36:11.732Z [volumes-ui]   File "<frozen importlib._bootstrap_external>", line 848, in exec_module
2024-08-13T17:36:11.732Z [volumes-ui]   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
2024-08-13T17:36:11.732Z [volumes-ui]   File "/entrypoint.py", line 4, in <module>
2024-08-13T17:36:11.732Z [volumes-ui]     from apps import default
2024-08-13T17:36:11.732Z [volumes-ui]   File "/apps/default/__init__.py", line 3, in <module>
2024-08-13T17:36:11.732Z [volumes-ui]     from kubeflow.kubeflow.crud_backend import config, logging
2024-08-13T17:36:11.732Z [volumes-ui] ModuleNotFoundError: No module named 'kubeflow'
2024-08-13T17:36:11.732Z [volumes-ui] [2024-08-13 17:36:11 +0000] [17] [INFO] Worker exiting (pid: 17)
2024-08-13T17:36:11.749Z [volumes-ui] [2024-08-13 17:36:11 +0000] [15] [ERROR] Worker (pid:17) exited with code 3
2024-08-13T17:36:11.750Z [volumes-ui] [2024-08-13 17:36:11 +0000] [15] [ERROR] Shutting down: Master
2024-08-13T17:36:11.750Z [volumes-ui] [2024-08-13 17:36:11 +0000] [15] [ERROR] Reason: Worker failed to boot.
2024-08-13T17:36:11.768Z [pebble] Change 1 task (Start service "volumes-ui") failed: cannot start service: exited quickly with code 3
2024-08-13T17:36:11.777Z [pebble] GET /v1/changes/1/wait 255.472656ms 200
2024-08-13T17:36:11.777Z [pebble] Started default services with change 1.
```

File: volumes-web-app/rockcraft.yaml [hunk 1]
```diff
@@ -1,96 +1,106 @@
...
       - python3-venv
       - python3-setuptools
       - python3-pip
     override-build: |
       pip3 install $CRAFT_STAGE/kubeflow-1.1-py3-none-any.whl
-      cd components/crud-web-apps/volumes/backend
+      cd $CRAFT_PART_SRC_WORK
       cp -r $CRAFT_STAGE/default apps/default/static/
       cp -r apps $CRAFT_STAGE/
       cp entrypoint.py $CRAFT_STAGE
```
Comment (kimwnasptd) [inline volumes-web-app/rockcraft.yaml:106]: Can we have the code to be almost identical to the one in the jupyter-web-app? Those 2 rocks should be almost identical
https://github.com/canonical/kubeflow-rocks/blob/main/jupyter-web-app/rockcraft.yaml#L122-L126

File: volumes-web-app/rockcraft.yaml [hunk 2]
```diff
@@ -1,120 +1,133 @@
...
 license: Apache-2.0
-base: ubuntu:20.04
+base: ubuntu@20.04
+platforms:
+  amd64:
+run-user: _daemon_
 
 services:
-  serve:
+  volumes-ui:
```
Comment (NohaIhab) [inline volumes-web-app/rockcraft.yaml:16]: the service name should be the same as the one in the [charm](https://github.com/canonical/kubeflow-volumes-operator/blob/track/1.9/src/charm.py#L120)
```suggestion
  kubeflow-volumes:
```

File: volumes-web-app/rockcraft.yaml [hunk 3]
```diff
@@ -1,120 +1,133 @@
...
 summary: Volumes Web App
 description: |
   This web app is responsible for allowing the user to manipulate PVCs in their Kubeflow
   cluster. To achieve this it provides a user friendly way to handle the lifecycle of PVC
   objects.
-version: v1.7.0_1
+version: v1.9.0
 license: Apache-2.0
-base: ubuntu:20.04
+base: ubuntu@20.04
```
Comment (NohaIhab) [inline volumes-web-app/rockcraft.yaml:10]:
```suggestion
base: ubuntu@22.04
```
for 1.9 rocks, we are using base 22.04

File: volumes-web-app/rockcraft.yaml [hunk 4]
```diff
@@ -1,120 +1,133 @@
+# Based on https://github.com/kubeflow/kubeflow/blob/v1.9.0/components/crud-web-apps/volumes/Dockerfile
 name: volumes-web-app
 summary: Volumes Web App
 description: |
   This web app is responsible for allowing the user to manipulate PVCs in their Kubeflow
   cluster. To achieve this it provides a user friendly way to handle the lifecycle of PVC
   objects.
-version: v1.7.0_1
+version: v1.9.0
```
Comment (NohaIhab) [inline volumes-web-app/rockcraft.yaml:8]:
```suggestion
version: 1.9.0
```
nit: to be consistent with the other rocks' version

**Review by NohaIhab — COMMENTED** [review]
I've tested the rock with the charm, I needed to make the following change in the charm's `pebble_components.py` for it to function correctly:
```
--- a/src/components/pebble_components.py
+++ b/src/components/pebble_components.py
@@ -28,7 +28,7 @@ class KubeflowVolumesPebbleService(PebbleServiceComponent):
             {
                 "services": {
                     self.service_name: {
-                        "override": "replace",
+                        "override": "merge",
                         "summary": "entry point for kubeflow-volumes",
                         "command": "/bin/bash -c 'gunicorn -w 3 --bind 0.0.0.0:5000 --access-logfile - entrypoint:app'",  # noqa: E501
                         "startup": "enabled",
```
however, this is expected as we are doing the same for `jupyter-web-app` [charm](https://github.com/canonical/notebook-operators/blob/track/1.9/charms/jupyter-ui/src/charm.py#L187).
Just a small comment on the rock `version`

**Review by NohaIhab — APPROVED** [review]
LGTM, thx @TakoB222

**Pull request merged**